src/java/org/apache/cassandra/index/IndexStatusManager.java

```diff
@@ -183,35 +212,81 @@ public synchronized void propagateLocalIndexStatus(String keyspace, String index
     {
         try
         {
-            Map<String, Index.Status> states = peerIndexStatus.computeIfAbsent(FBUtilities.getBroadcastAddressAndPort(),
+            Map<String, Index.Status> statusMap = peerIndexStatus.computeIfAbsent(FBUtilities.getBroadcastAddressAndPort(),
                                                                                k -> new HashMap<>());
             String keyspaceIndex = identifier(keyspace, index);
 
             if (status == Index.Status.DROPPED)
-                states.remove(keyspaceIndex);
+                statusMap.remove(keyspaceIndex);
             else
-                states.put(keyspaceIndex, status);
+                statusMap.put(keyspaceIndex, status);
 
             // Don't try and propagate if the gossiper isn't enabled. This is primarily for tests where the
             // Gossiper has not been started. If we attempt to propagate when not started an exception is
             // logged and this causes a number of dtests to fail.
             if (Gossiper.instance.isEnabled())
             {
-                String newStatus = JsonUtils.JSON_OBJECT_MAPPER.writeValueAsString(states);
+                // Versions 5.0.0 through 5.0.2 use a much more bloated format that duplicates keyspace names
+                // and writes full status names instead of their numeric codes. If the minimum cluster version is
+                // unknown or one of those 3 versions, continue to propagate the old format.
+                CassandraVersion minVersion = Gossiper.instance.getMinVersion(1, TimeUnit.SECONDS);
+                String newSerializedStatusMap = shouldWriteLegacyStatusFormat(minVersion) ? JsonUtils.writeAsJsonString(statusMap) 
+                                                                                          : toSerializedFormat(statusMap);
+
                 statusPropagationExecutor.submit(() -> {
                     // schedule gossiper update asynchronously to avoid potential deadlock when another thread is holding
                     // gossiper taskLock.
-                    VersionedValue value = StorageService.instance.valueFactory.indexStatus(newStatus);
+                    VersionedValue value = StorageService.instance.valueFactory.indexStatus(newSerializedStatusMap);
                     Gossiper.instance.addLocalApplicationState(ApplicationState.INDEX_STATUS, value);
                 });
             }
         }
-        catch (Throwable e)
+        catch (Exception e)
         {
             logger.warn("Unable to propagate index status: {}", e.getMessage());
         }
     }
 
+    private static boolean shouldWriteLegacyStatusFormat(CassandraVersion minVersion)
+    {
+        return minVersion == null || (minVersion.major == 5 && minVersion.minor == 0 && minVersion.patch < 3);
```

Comment:
We just released 5.0.3, so this will have to be bumped...

Reply:
this looks fine, but wonder why we check each field rather than doing what we do in other places?

```
static final CassandraVersion SUPPORTS_RETRY = new CassandraVersion("5.0.0-alpha2.SNAPSHOT");
...
if (remoteVersion.compareTo(SUPPORTS_RETRY) >= 0)
  ...
```

For a production release you are fine, but for "snapshot" releases its a bit more iffy... `5.0.3-SNAPSHOT` is before `5.0.3` but after `5.0.2`.  So if you have a running snapshot cluster and upgrade this check will say we are fine but we actually are not.  Do we need to optimize for this case no, more just calling out these possibilities.

Reply:
The snapshot concern is fair, although it's probably reasonable to assume that once this commit lands, we just tell people to use the latest snapshot. My real concern is that 5.0.3 is being released soon, and I'm not sure if this is going to have to be `< 5.0.4`...